### PR TITLE
New: `lines-around-directive` rule (fixes #6069)

### LINKED ISSUE #6069: Rule Proposal: lines-around-directive
From [requirePaddingNewLinesAfterUseStrict] and [disallowPaddingNewLinesAfterUseStrict].

This rule will require/disallow blank lines before/after a chunk of directive prologues.

```js
{
    "lines-around-directive": ["error", "always" or "never"],
    // or
    "lines-around-directive": ["error", {
        "after": "always" or "never",
        "before": "always" or "never"
    }]
}
```

**Valid:**

```js
/*eslint lines-around-directive: ["error", "always"]*/

// header comment?

"use strict";

foo();

function foo() {
    "use strict";  // ignore at the head of block. this is warned by padded-blocks.
    "use asm";     // ignore between directives.

    foo();
}
```

```js
/*eslint lines-around-directive: ["error", "never"]*/

// header comment?
"use strict";
foo();

function foo() {

    "use strict";  // ignore at the head of block. this is warned by padded-blocks.
    "use asm";     // ignore between directives.
    foo();
}
```

**Invalid:**

```js
/*eslint lines-around-directive: ["error", "always"]*/

// header comment?
"use strict";
foo();

function foo() {
    "use strict";
    foo();
}
```

```js
/*eslint lines-around-directive: ["error", "never"]*/

// header comment?

"use strict";

foo();

function foo() {
    "use strict";

    foo();
}
```

[requirePaddingNewLinesAfterUseStrict]: http://jscs.info/rule/requirePaddingNewLinesAfterUseStrict
[disallowPaddingNewLinesAfterUseStrict]: http://jscs.info/rule/disallowPaddingNewLinesAfterUseStrict

### PR DESCRIPTION
**Pull request by kaicataldo**

### CODE REVIEW

**Comment by mysticatea** [comment]
Thank you for implementation!

My intention was that this rule handles directive prologues (not restrict `"use strict"`).

> A Directive Prologue is the longest sequence of ExpressionStatement productions occurring as the initial StatementListItem or ModuleItem productions of a FunctionBody, a ScriptBody, or a ModuleBody and where each ExpressionStatement in the sequence consists entirely of a StringLiteral token followed by a semicolon. The semicolon may appear explicitly or may be inserted by automatic semicolon insertion. A Directive Prologue may be an empty sequence.
> http://www.ecma-international.org/ecma-262/7.0/#directive-prologue

A directive prologue might be multiline:

    "use asm"
    "use strict"

**Comment by kaicataldo** [comment]
Ah, okay. So to check each directive in the prologue I could maybe do something like this (to borrow [this implementation](https://github.com/eslint/eslint/blob/master/lib/rules/strict.js#L35))

```js
function getDirectives(statements) {
    const directives = [];

    for (let i = 0; i < statements.length; i++) {
        const statement = statements[i];

        if (
            statement.type === "ExpressionStatement" &&
            statement.expression.type === "Literal"
        ) {
            directives[i] = statement;
        } else {
            break;
        }
    }

    return directives;
}
```

**Comment by mysticatea** [comment]
Thank you.
I think good to move the logic to `ast-utils`.

By the way, `directives[i] = statement` seems a bug. It should be `directives.push(statement)` I think.

**Comment by kaicataldo** [comment]
You're right! Do you mean adding a `getDirectives` utility to `ast-utils`? Thanks!

**Comment by mysticatea** [comment]
Yes!

**Comment by kaicataldo** [comment]
Got it working for multiple directives and then realized I wasn't sure if we want to check each directive individually or if we want to check around the entire directive prologue. Any thoughts on this @mysticatea?

Now that I know we're checking all directives, I think it makes sense to have an option to enforce blank newlines around the whole prologue as well as each individual directive. This wasn't discussed in the accepted issue though, so maybe we should talk about it there.

**Comment by mysticatea** [comment]
Actually, my first post #6069 says "ignores between directives".
My intention was that the rule checks blank lines around a chunk of directives. :wink:

**Comment by kaicataldo** [comment]
Ah, so it does...Missed that somehow. Okay, thanks! Will update accordingly 👍

**Comment by kaicataldo** [comment]
Still have to update the docs

**Comment by kaicataldo** [comment]
Updated with new docs. Please let me know what you think, @mysticatea

**Comment by mysticatea** [comment]
Wow, there are super many tests!
Looks great to me.

File: lib/ast-utils.js [hunk 1]
```diff
@@ -680,5 +680,46 @@ module.exports = {
...
+
+        for (const statement of statements) {
+            if (
+                statement.type === "ExpressionStatement" &&
+                statement.expression.type === "Literal"
+            ) {
+                directives.push(statement);
+            } else {
+                if (!directives.length) {
+                    return null;
```
Comment (mysticatea) [inline lib/ast-utils.js]: Personally, I prefer an empty array instead of `null`.

(currently, this returns an empty array if the function is empty?)

Comment (kaicataldo) [inline lib/ast-utils.js]: Would your preference be for it to return an empty array for the early return as well? My goal was to make it consistent for when it doesn't find directives.

Comment (mysticatea) [inline lib/ast-utils.js]: My preference is that the function always returns an array. We can check the existence of directive prologues by the `length` of the result (or by iteration of the result).

Also, current implementation seems inconsistent.

```js
// In empty function case, returns `[]`
function foo() {
}

// In non-strict function case, returns `null`
function foo() {
    bar()
}
```

Comment (kaicataldo) [inline lib/ast-utils.js]: It should return `null` in both cases (https://github.com/eslint/eslint/pull/6998/files#diff-6198b73cda8ab9ac49ed872ead7fb268R716), but I can definitely make it always return an array :+1:

**Comment by kaicataldo** [comment]
Updated once more

**Comment by mysticatea** [comment]
Great, thank you very much!
LGTM.

File: docs/rules/lines-around-directive.md [hunk 2]
```diff
@@ -0,0 +1,320 @@
+# Require or disallow newlines around directives (lines-around-directive)
```
Comment (nzakas) [inline docs/rules/lines-around-directive.md]: I think we are doing lowercase first word now.

**Comment by nzakas** [comment]
Lgtm, just one tiny thing on the docs. Really nice job on this -- I especially love the intro on the docs. Very well-written. 💯

**Comment by vitorbal** [comment]
Super awesome, @kaicataldo!! Great job! 🎉  🎉

**Pull request merged**